File: src/main/java/com/sf_lolitahag/action/motion/Ball.java
/*
Copyright (C) 2016 Koji Wakamiya
Copyright (C) 2016 Masanori Kojima
Released under the MIT license
http://opensource.org/licenses/mit-license.php
*/
package com.sf_lolitahag.action.motion;

import com.sf_lolitahag.Utils;
import com.sf_lolitahag.action.pitchs.Fast;
import com.sf_lolitahag.action.pitchs.IPitch;
import com.sf_lolitahag.action.pitchs.Liner;
import com.sf_lolitahag.action.pitchs.Makyu;
import com.sf_lolitahag.action.pitchs.Straight;
import java.awt.Image;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import javax.swing.Timer;
import org.ehcache.config.builders.CacheConfigurationBuilder;
import org.ehcache.config.builders.ResourcePoolsBuilder;

/**
 * ボール実装
 */
public class Ball extends AbstractMotion {

  private static final int AXIS_X = 475;
  private static final int AXIS_Y = 245;
  private static final int PAINT_INTERVAL = 40;
  private static final int PAINT_INTERVAL_SPIN = 25;
  private static final int HIT_ZONE_START = 560;
  private static final int HIT_ZONE_END = 660;
  private static final List<String> BALL = Collections.singletonList("ball01");
  private static final List<String> BALL_SPIN = Arrays.asList("spin01", "spin02");
  private static final List<String> IMAGE_LIST = Arrays.asList("ball01", "spin01", "spin02");
  private int index = 0;
  private boolean isHitZone;
  private final Callback callback;
  private IPitch pitch;
  private final Timer spinTimer = new Timer(PAINT_INTERVAL_SPIN, (e) -> updateSpin());
  private final Timer paintTimer = new Timer(PAINT_INTERVAL, (e) -> updateBallView());

  public Ball(Callback callback) {
    this.callback = callback;

    cache = CACHE_MANAGER.createCache(Ball.class.getSimpleName(), CacheConfigurationBuilder
        .newCacheConfigurationBuilder(String.class, Image.class, ResourcePoolsBuilder.heap(100))
        .build());
    IMAGE_LIST.forEach(name -> cache.put(name, Utils.getImageFromResources(getClass(), name)));

    init();
  }

  public void startPitch() {
    getPitchRand();
    isShow = true;
    paintTimer.start();
  }

  public boolean isHit() {
    if (isHitZone) {
      pitch = new Liner();
    }

    return isHitZone;
  }

  private void init() {
    axisX = AXIS_X;
    axisY = AXIS_Y;
    fileName = BALL.get(index);
    isShow = false;
    isHitZone = false;
  }

  private void getPitchRand() {
    switch (BallType.getBallType()) {
      case STRAIGHT:
        pitch = new Straight();
        break;
      case FAST:
        pitch = new Fast();
        break;
      case MAKYU:
        pitch = new Makyu();
        break;
    }
  }

  private void updateBallView() {
    int updateX = pitch.getUpdateX(axisX);
    axisX += updateX;
    int updateY = pitch.getUpdateY(axisY);
    axisY += updateY;
    isHitZone = axisY >= HIT_ZONE_START && axisY <= HIT_ZONE_END;

    if (pitch.isSpin()) {
      spinTimer.start();
    } else {
      spinTimer.stop();
      index = 0;
      fileName = BALL.get(index);
    }

    checkFinish(updateY);
  }

  private void checkFinish(int updateY) {
    int axisY = this.axisY + updateY;
    if (axisY < AXIS_Y || axisY >= 800) {
      paintTimer.stop();
      callback.onFinishPitch(axisY >= 800);
      init();
    }
  }

  private void updateSpin() {
    if (index == 0) {
      index = 1;
    } else if (index == 1) {
      index = 0;
    }
    fileName = BALL_SPIN.get(index);
  }

  private enum BallType {
    STRAIGHT,
    FAST,
    MAKYU;

    public static BallType getBallType() {
      int index = (int) (Math.random() * BallType.values().length);
      return getBallTypeByIndex(index);
    }

    private static BallType getBallTypeByIndex(int index) {
      List<BallType> objects = Arrays.asList(BallType.values());
      return objects.get(index);
    }
  }

  public interface Callback {

    void onFinishPitch(boolean isWin);
  }
}

File: src/main/kotlin/com/sf_lolitahag/motion/Ball.kt
/*
Copyright (C) 2016 Koji Wakamiya
Copyright (C) 2016 Masanori Kojima
Released under the MIT license
http://opensource.org/licenses/mit-license.php
*/
package com.sf_lolitahag.motion

import com.sf_lolitahag.Utils
import com.sf_lolitahag.pitchs.*
import org.ehcache.config.builders.CacheConfigurationBuilder
import org.ehcache.config.builders.ResourcePoolsBuilder
import java.awt.Image
import javax.swing.Timer

class Ball(private val callback: Callback) : AbstractMotion() {

    companion object {
        private const val AXIS_X = 475
        private const val AXIS_Y = 245
        private const val PAINT_INTERVAL = 40
        private const val PAINT_INTERVAL_SPIN = 25
        private const val HIT_ZONE_START = 560
        private const val HIT_ZONE_END = 660
        private val BALL = listOf("ball01")
        private val BALL_SPIN = listOf("spin01", "spin02")
        private val IMAGE_LIST = listOf("ball01", "spin01", "spin02")
    }

    private var index = 0
    private var isHitZone: Boolean = false
    private var pitch: IPitch? = null
    private val spinTimer = Timer(PAINT_INTERVAL_SPIN) { updateSpin() }
    private val paintTimer = Timer(PAINT_INTERVAL) { updateBallView() }

    init {
        cache = cacheManager.createCache(Ball::class.java.simpleName, CacheConfigurationBuilder
                .newCacheConfigurationBuilder(String::class.java, Image::class.java, ResourcePoolsBuilder.heap(100))
                .build())
        IMAGE_LIST.forEach { name -> cache.put(name, Utils.getImageFromResources(javaClass, name)) }

        init()
    }

    val isHit: Boolean
        get() {
            if (isHitZone) {
                pitch = Liner()
            }

            return isHitZone
        }

    fun startPitch() {
        getPitchRand()
        isShow = true
        paintTimer.start()
    }

    private fun init() {
        axisX = AXIS_X
        axisY = AXIS_Y
        fileName = BALL[index]
        isShow = false
        isHitZone = false
    }

    private fun getPitchRand() {
        pitch = when (BallType.ballType) {
            BallType.STRAIGHT -> Straight()
            BallType.FAST -> Fast()
            BallType.MAKYU -> Makyu()
        }
    }

    private fun updateBallView() {
        val updateX = pitch!!.getUpdateX(axisX)
        axisX += updateX
        val updateY = pitch!!.getUpdateY(axisY)
        axisY += updateY
        isHitZone = axisY in HIT_ZONE_START..HIT_ZONE_END

        if (pitch!!.isSpin) {
            spinTimer.start()
        } else {
            spinTimer.stop()
            index = 0
            fileName = BALL[index]
        }

        checkFinish(updateY)
    }

    private fun checkFinish(updateY: Int) {
        val axisY = this.axisY + updateY
        if (axisY < AXIS_Y || axisY >= 800) {
            paintTimer.stop()
            callback.onFinishPitch(axisY >= 800)
            init()
        }
    }

    private fun updateSpin() {
        if (index == 0) {
            index = 1
        } else if (index == 1) {
            index = 0
        }
        fileName = BALL_SPIN[index]
    }

    private enum class BallType {
        STRAIGHT,
        FAST,
        MAKYU;


        companion object {

            val ballType: BallType
                get() {
                    val index = (Math.random() * BallType.values().size).toInt()
                    return getBallTypeByIndex(index)
                }

            private fun getBallTypeByIndex(index: Int): BallType {
                val objects = listOf(*BallType.values())
                return objects[index]
            }
        }
    }

    interface Callback {

        fun onFinishPitch(isWin: Boolean)
    }
}
